### PR TITLE
custom and functioning jwt authentication

### PR DESCRIPTION
**Pull request by beetai**
😭 🔫

### CODE REVIEW

File: frontend/src/redux/reducers/index.tsx [hunk 1]
```diff
@@ -11,6 +13,8 @@ export default combineReducers({
   users: getUsers,
   currentUser: getCurrentUser,
   visualizedCourses,
+  auth: authReducer,
+  error: errorReducer,
```
Comment (DSep) [inline frontend/src/redux/reducers/index.tsx:17]: Cool. Let's use this for other stuff.

**Comment by beetai** [comment]
(react) routing are 10000% broken, in relation to auth, but the redux and backend should all be good

File: frontend/src/components/ProtectedRoute.tsx [hunk 2]
```diff
@@ -4,12 +4,12 @@ import { Route, Redirect } from 'react-router-dom';
...
 
 function ProtectedRoute(props: any) {
-  const { currentUser, component: Component, ...rest } = props;
+  const { auth, component: Component, ...rest } = props;
   return (
     <Route
       {...rest}
       render={(props) => (
-        currentUser.name
+        auth.user
```
Comment (beetai) [inline frontend/src/components/ProtectedRoute.tsx]: change to `auth.isAuthenticated`

File: frontend/src/components/SignupForm.tsx [hunk 3]
```diff
@@ -14,42 +19,98 @@ const useStyles = makeStyles((theme: Theme) => createStyles({
...
+      email,
+      password,
+    };
+
+    register(newUser);
+  };
+
+  useEffect(() => {
+    if (error.id === REGISTER_FAIL) {
+      setErrorMsg(error.msg.msg);
```
Comment (DSep) [inline frontend/src/components/SignupForm.tsx]: hmmmmm

File: frontend/src/components/SignupForm.tsx [hunk 4]
```diff
@@ -14,42 +19,98 @@ const useStyles = makeStyles((theme: Theme) => createStyles({
...
-      <Button variant="contained" className={classes.loginButton}>Sign up</Button>
+      <Button variant="contained" className={classes.loginButton} onClick={handleSubmit}>Sign up</Button>
     </form>
   );
 };
 
-export default SignupForm;
+const mapStateToProps = (state: any) => ({
+  isAuthenticated: state.auth.isAuthenticated,
+  error: state.error,
```
Comment (DSep) [inline frontend/src/components/SignupForm.tsx:113]: seems sus

File: frontend/src/components/ProtectedRoute.tsx (hunk 2 above)
Comment (DSep) [inline frontend/src/components/ProtectedRoute.tsx]: I thought there was an auth.isAuthenticated? Use that maybe

File: frontend/src/pages/AccountForm.tsx [hunk 5]
```diff
@@ -31,7 +32,8 @@ function AccountForm(props: any) {
 
   useEffect(() => {
     // TODO: make this a specific get current user action
```
Comment (DSep) [inline frontend/src/pages/AccountForm.tsx]: delete comments pls

File: frontend/src/pages/AccountForm.tsx [hunk 6]
```diff
@@ -141,10 +143,7 @@ function AccountForm(props: any) {
...
 
-const mapStateToProps = (state: any) => {
-  const { currentUser } = state;
-  return { currentUser };
-};
+const mapStateToProps = (state: any) => ({ currentUser: state.auth.user });
 
-export default connect(mapStateToProps, { getUsers, updateUser })(AccountForm);
+export default connect(mapStateToProps, { loadUser, updateUser })(AccountForm);
 // TODO: change get users to get specific user. this is placeholder before we implement new action
```
Comment (DSep) [inline frontend/src/pages/AccountForm.tsx]: comment

File: frontend/src/pages/Login.tsx [hunk 7]
```diff
@@ -23,7 +23,7 @@ const Login = (props: any) => {
   const classes = useStyles();
   console.log(users);
   useEffect(() => {
-    getUsers();
+    // getUsers();
```
Comment (DSep) [inline frontend/src/pages/Login.tsx:26]: comment

File: frontend/src/redux/actions/auth.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+import {
+  USER_LOADING,
+  USER_LOADED,
+  AUTH_ERROR,
+  LOGIN_SUCCESS,
+  LOGIN_FAIL,
+  LOGOUT_SUCCESS,
+  REGISTER_SUCCESS,
+  REGISTER_FAIL, GET_ERRORS,
```
Comment (DSep) [inline frontend/src/redux/actions/auth.ts]: remove get errors

**Review by DSep — CHANGES_REQUESTED** [review]
Just a few things

**Review by yqj2k — APPROVED** [review]
robusstttttttttttttttt

File: frontend/src/components/SignupForm.tsx (hunk 3 above)
Comment (beetai) [reply frontend/src/components/SignupForm.tsx]: NAAAAAAH dude nothing wrong here

File: frontend/src/components/ProtectedRoute.tsx (hunk 2 above)
Comment (beetai) [reply frontend/src/components/ProtectedRoute.tsx]: yup, gonna do that

**Review by yqj2k — APPROVED** [review]
now we're robust

**Comment by beetai** [comment]
y'all don't appreciate msg.msg smh

File: server/routes/auth.js [hunk 9]
```diff
@@ -12,7 +12,7 @@ router.post('/', async (req, res, next) => {
 
   // validation
   if (!email || !password) {
-    return res.status(400).json({ msg: 'Please enter all required fields' });
+    return res.status(400).json('Please enter all required fields');
```
Comment (DSep) [inline server/routes/auth.js:15]:
```suggestion
    const msg= {};
    for (let i = 0; i < Math.floor(Math.random()*10000); i++) {
      msg.msg = {};
    }
    msg.msg = 'Please enter all required fields';
    return res.status(400).json(msg);
```
@beetai @sam-ip @baconandchips

Comment (beetai) [reply server/routes/auth.js:15]: only galaxy brain would do this

**Review by DSep — APPROVED** [review]
LGTM

**Review by baconandchips — APPROVED** [review]
LGTM

File: frontend/src/pages/AccountForm.tsx [hunk 10]
```diff
@@ -12,15 +12,16 @@ import TextField from '@material-ui/core/TextField';
...
+  loadUser: Function,
   updateUser: Function,
 }
 
 function AccountForm(props: any) {
-  const { currentUser, getUsers, updateUser } : Props = props;
+  const { currentUser, loadUser, updateUser } : Props = props;
   const [name, setName] = useState(currentUser.name);
   const [email, setEmail] = useState(currentUser.email);
   const [major, setMajor] = useState(currentUser.major);
```
Comment (baconandchips) [inline frontend/src/pages/AccountForm.tsx:27]: major, when clicked into after account creation (where its initial value is undefined or empty string), gives a big red error.

"Warning: A component is changing an uncontrolled input of type text to be controlled. Input elements should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://fb.me/react-controlled-components"

**Review by yqj2k — APPROVED** [review]

**Pull request merged**